### ISSUE TITLE
RFC: `z.envbool()`

### ISSUE THREAD

**Issue by colinhacks** [colinhacks/zod#3906]
An customizable, first-class way to coerce string inputs to `boolean`, inspired by common environment variable conventions. 

```
z.envbool()
```

Default behavior:

- Returns `true` for `["true", "1", "yes", "on"]` (case insensitive)
- Returns `false` for `["false", "0", "no", "off"]` (case insensitive)
- Throws otherwise

The choice to throw on unrecognized values is intentional. This API will be used for parsing potentially sensitive environment variables, so this API intentionally avoids any non-explicit defaults (e.g. defaulting to `false`) to avoid potentially unexpected behavior.

To specify a value to be returned when an uexpected input is encountered, users can use the pre-existing `.catch()` method.

```ts
z.envbool().catch(false)
```

To customize the set of truthy and falsy strings:

```ts
z.envbool({
  true: ["yes", "on", "enabled", "y"],
  false: ["no", "off", "disabled", "n"]
})
```

To customize the set of truthy and falsy strings:

```ts
z.envbool({
  true: ["yes", "on", "enabled", "y"],
  false: ["no", "off", "disabled", "n"]
})
```

To make it case sensitive:

```ts
z.envbool({
  case: "sensistive" // default "insensitive"
})
```

**Comment by nathggns** [comment]
Love it. This would replace something we've had to build ourselves, and far more robustly too.

**Comment by crisshaker** [comment]
Ohh so glad this is going to get native support. Curious though, having 'env' in the name seems a little too specific/suggesting? I know there's nothing preventing using it any other way but just curious if it's really intended.

**Comment by kevinbailey25** [comment]
This is handy, we have something to had to build to solve this. Specifically when handling checkboxes on posted forms. We'd do a transform to see if the value was === "TRUE" or whatever value our checkbox was. 

I'm not sure how I feel about the "envbool" name tho.

**Comment by 4lve** [comment]
Good idea, though it might make it clearer if you named it something like 'commonbool', 'humanbool', 'stringbool'

**Comment by quantizor** [comment]
normalizedBool might be easier to immediately understand

**Comment by Vinlock** [comment]
Would be interesting if it was something like
```ts
z.coerce.boolean({
  acceptedTruthyValues: [...],
  acceptedFalsyValues: [...]
})
```

**Comment by yousefelgoharyx** [comment]
aliasedBool is so much better to understand

**Comment by scotttrinh** [comment]
As far as this proposal is concerned, as-is, I agree that "env" is a little too suggestive since I think the most common case I've seen users hit this snag is in search parameter parsing, which wants an even subtly different set of defaults (like `""` being truthy...). What if we exposed just the more manual API with no "default" and exported some common configurations like:

```ts
import { z } from "zod";

// The current defaults
z.boolFrom(z.ENV_BOOL_FROM);

// URLSearchParams value
z.boolFrom(z.SEARCH_PARAMS_BOOL_FROM);
```

I don't actually _like_ that (maybe separate methods on `boolFrom`?) but I think that avoids this being hard to intuit.

---

I wonder if there is a more general solution to "map from some input domain into some enumerated domain"? I guess that's just union-of-pipe, right? But, that would need to be combined with some kind of easy-to-use case-insensitive literal schema to approach the same ergonomics.

```ts
import { z } from "zod";

z.union([
  z.stringOf(["y", "true", "yes"]).pipe(z.true()),
  z.stringOf(["n", "false", "no"]).pipe(z.false()),
]);
```

**Comment by Ashwin-Mahadevan** [comment]
This proposal feels like it should be in the `z.coerce` api. I agree with @Vinlock, although I'd prefer this syntax:

```ts

import { z } from "zod";

const first = z.coerce.envbool();

const second = z.coerce.envbool({
    truthy: ["true", "1", "..."],
    falsy: ["false", "0", "..."],
    case: "sensitive",
});
```

**Comment by soulchild** [comment]
boolish? 😄

**Comment by joshuayoes** [comment]
This feels like this should be a part of the [String collection](https://zod.dev/?id=strings)

```ts
z.string().bool()
```

Especially since there are other string parsed types like [Datetime](https://zod.dev/?id=datetimes) which can be passed options like this:

```ts
const datetime = z.string().datetime({ offset: true });
```

This would feel pretty natural since these extra string validations already can be passed options.

```ts
z.string().bool({
  true: ["y"],
  false: ["n"]
})
```

**Comment by alif-devcon** [comment]
need this.

**Comment by nomis6432** [comment]
Hey, I was wondering if alternatively, it would be a solution to add custom coercion rules globally? F.e. using this syntax:

```ts
z.coerce.boolean().parse('false'); // true

z.coerce.boolean.setParser((val: any): boolean => {
  if (typeof val === 'string') {
    if (['1', 'true'].includes(val.toLowerCase())) return true;
    if (['0', 'false'].includes(val.toLowerCase())) return false;
    throw new Error('Invalid boolean string');
  }
  return Boolean(val);
});

z.coerce.boolean().parse('false'); // false
```

This would remove boilerplate code needed to always define your coercion rules each time you use `z.coerce.boolean()`

In some cases it might not be desirable to modify the global coercion state (f.e. big monorepos). In such cases, it might be useful to have a syntax like this:

```ts
zz = z.getCustomCoerce()
zz.boolean.setParser((val: any): boolean => {
  if (typeof val === 'string') {
    if (['1', 'true'].includes(val.toLowerCase())) return true;
    if (['0', 'false'].includes(val.toLowerCase())) return false;
    throw new Error('Invalid boolean string');
  }
  return Boolean(val);
});
zz.boolean().parse("false") // false
z.coerce.boolean().parse("false") // true
```

**Comment by Jay-Karia** [comment]
Implemented `z.envbool()`
https://github.com/colinhacks/zod/pull/4054

Open for suggestions

**Comment by soulchild** [comment]
I can imagine this being useful for other things than coercing strings from environment variables to boolean values. Hence I think the name should be more generic and not necessarily contain `env`.

**Comment by Jay-Karia** [comment]
Let's vote for the name:

- 👍 `z.envbool()`
- 😄 `z.boolish()`
- 🎉 `z.stringbool()`
- ❤️ `z.stringAsBool()`
- 🚀 `z.string().bool()`
- 👀 `z.bool()`

**Comment by mlkirkpatrick** [comment]
Can I request:
- Returns `true`: `"enable"|"enabled"`
- Returns `false`: `"disable"|"disabled"`

The documentation currently mentions `"enable"` and `"disabled`"  which mixes verb and past participle/adjective. 😄 I would probably use the `-ed` ending, but I could definitely see people using either form.

**Comment by colinhacks** [comment]
Landed as `z.stringbool()` in Zod 4: https://zod.dev/v4?id=stringbool
